Add SMLSL, SQRSHRN and SRSHR (Vector) cpu instructions, nits

gdkchan opened the pull request:
Since SQRSHRN is a saturate + narrow operation (like SQXTN and friends) I moved the code used by this instruction to AVectorSimdHelper (where the shared methods are located) and modified it to accept an Action argument used to emit the extra round + shift operation needed by this instruction. I also modified it to avoid emiting a IL conversion instruction (Conv_I8/U8) by using a long constant for the value ranges. Not sure if this is the best way to arrange stuff as there are also other instruction that needs, for example, saturation without narrowing.

ptal @LDj3SNuD (I ran the tests and they all pass so at least the SQXTN/UQXTN instructions should be working properly with those changes).

LDj3SNuD commented:
The instructions `Sqxtn, Uqxtn, Sqxtun (S, V)` have been successfully re-tested.

LDj3SNuD commented:
I agree with your solution (I did not dig into it).

I have two further notes to make:
1) The tables of `Sshr_V, Ssra_V, Ushr_V, Usra_V` should be updated with the same pattern.
2) It would be useful to coagulate these 2 private methods:
https://github.com/gdkchan/Ryujinx/blob/5245960880320089e15f15643be041010c6cc54f/ChocolArm64/Instruction/AInstEmitSimdShift.cs#L295
https://github.com/gdkchan/Ryujinx/blob/5245960880320089e15f15643be041010c6cc54f/ChocolArm64/Instruction/AInstEmitSimdShift.cs#L329
since they share the same operations (currently one method does not allow accumulation, while the other does not allow rounding):
```
SRSRA (srsra_advsimd.html)
SSRA (ssra_advsimd.html)
SRSHR (srshr_advsimd.html)
SSHR (sshr_advsimd.html)
And the 4 unsigned
```

LDj3SNuD commented on ChocolArm64/Instruction/AInstEmitSimdShift.cs:
```diff
@@ -115,7 +143,7 @@ public static void Sshr_V(AILEmitterCtx Context)
             AOpCodeSimdShImm Op = (AOpCodeSimdShImm)Context.CurrOp;
 
-            int Shift = (8 << (Op.Size + 1)) - Op.Imm;
+            int Shift = GetImmShr(Op);
```
remove ...

LDj3SNuD commented on ChocolArm64/Instruction/AInstEmitSimdShift.cs:
```diff
@@ -115,7 +143,7 @@ public static void Sshr_V(AILEmitterCtx Context)
-            int Shift = (8 << (Op.Size + 1)) - Op.Imm;
+            int Shift = GetImmShr(Op);
 
             EmitVectorShImmBinarySx(Context, () => Context.Emit(OpCodes.Shr), Shift);
```
... add

LDj3SNuD replied:
... add

gdkchan commented:
Sorry for the delay. I made all requested changes:

- Fixed encodings for `Sshr_V`, `Ssra_V`, `Ushr_V` and `Usra_V`.
- Additionally, also fixed encodings for `Sshr_S` and `Ushr_S`.
- Removed the `EmitVectorRoundShImmOp` method. Made `EmitVectorShImmOp` more flexible by adding a flags argument used to control the different variations of the immediate shift instructions. Now round uses it aswell.
- Removed the useless `Shift` temp var.

LDj3SNuD commented on ChocolArm64/AOpCodeTable.cs:
```diff
@@ -371,16 +371,22 @@ static AOpCodeTable()
-            SetA64("0x0011110>>>>xxx000001xxxxxxxxxx", AInstEmit.Sshr_V,        typeof(AOpCodeSimdShImm));
-            SetA64("0x0011110>>>>xxx000101xxxxxxxxxx", AInstEmit.Ssra_V,        typeof(AOpCodeSimdShImm));
+            SetA64("0101111101xxxxxx000001xxxxxxxxxx", AInstEmit.Sshr_S,        typeof(AOpCodeSimdShImm));
+            SetA64("0000111100>>>xxx000001xxxxxxxxxx", AInstEmit.Sshr_V,        typeof(AOpCodeSimdShImm));
```
"0x00111100>>>xxx000001xxxxxxxxxx"

LDj3SNuD commented on ChocolArm64/AOpCodeTable.cs:
```diff
@@ -371,16 +371,22 @@ static AOpCodeTable()
+            SetA64("0101111101xxxxxx000001xxxxxxxxxx", AInstEmit.Sshr_S,        typeof(AOpCodeSimdShImm));
+            SetA64("0000111100>>>xxx000001xxxxxxxxxx", AInstEmit.Sshr_V,        typeof(AOpCodeSimdShImm));
+            SetA64("0100111101xxxxxx000001xxxxxxxxxx", AInstEmit.Sshr_V,        typeof(AOpCodeSimdShImm));
+            SetA64("0000111100>>>xxx000101xxxxxxxxxx", AInstEmit.Ssra_V,        typeof(AOpCodeSimdShImm));
```
"0x00111100>>>xxx000101xxxxxxxxxx"

LDj3SNuD commented on ChocolArm64/AOpCodeTable.cs:
```diff
@@ -419,9 +425,11 @@ static AOpCodeTable()
-            SetA64("0x1011110>>>>xxx000001xxxxxxxxxx", AInstEmit.Ushr_V,        typeof(AOpCodeSimdShImm));
-            SetA64("0x1011110>>>>xxx000101xxxxxxxxxx", AInstEmit.Usra_V,        typeof(AOpCodeSimdShImm));
+            SetA64("0111111101xxxxxx000001xxxxxxxxxx", AInstEmit.Ushr_S,        typeof(AOpCodeSimdShImm));
+            SetA64("0010111100>>>xxx000001xxxxxxxxxx", AInstEmit.Ushr_V,        typeof(AOpCodeSimdShImm));
```
"0x10111100>>>xxx000001xxxxxxxxxx"

LDj3SNuD commented on ChocolArm64/AOpCodeTable.cs:
```diff
@@ -419,9 +425,11 @@ static AOpCodeTable()
+            SetA64("0111111101xxxxxx000001xxxxxxxxxx", AInstEmit.Ushr_S,        typeof(AOpCodeSimdShImm));
+            SetA64("0010111100>>>xxx000001xxxxxxxxxx", AInstEmit.Ushr_V,        typeof(AOpCodeSimdShImm));
+            SetA64("0110111101xxxxxx000001xxxxxxxxxx", AInstEmit.Ushr_V,        typeof(AOpCodeSimdShImm));
+            SetA64("0010111100>>>xxx000101xxxxxxxxxx", AInstEmit.Usra_V,        typeof(AOpCodeSimdShImm));
```
"0x10111100>>>xxx000101xxxxxxxxxx"

LDj3SNuD commented on ChocolArm64/Instruction/AInstEmitSimdShift.cs:
```diff
@@ -251,36 +273,64 @@ void EmitShift(OpCode ILOp)
+            for (int Index = 0; Index < Elems; Index++)
             {
-                if (Ternary)
+                if ((Flags & ShImmFlags.Ternary) != 0)
```
It would be helpful to keep this pre-calculable out of the for:
So, `bool Ternary = (Flags & ShImmFlags.Ternary) != 0;` under the line #316

LDj3SNuD commented on ChocolArm64/Instruction/AInstEmitSimdShift.cs:
```diff
@@ -251,36 +273,64 @@ void EmitShift(OpCode ILOp)
 
                 EmitVectorExtract(Context, Op.Rn, Index, Op.Size, Signed);
 
+                if ((Flags & ShImmFlags.Rounded) != 0)
```
And, `bool Rounded = (Flags & ShImmFlags.Rounded) != 0;` under the line #317

gdkchan commented:
Done.
